### PR TITLE
Adding support for cleaning up staging data when Gobblin is cancelled via the Azkaban UI

### PR DESCRIPTION
**Pull request by sahilTakiar**
Here is the first version. I still need to do some cleanup, add some comments, and write a test. Making this pull request to see what everyone thinks of the approach.

Azkaban only lets the cancel method run for 5 ms, before doing a hard kill on the process. This isn't enough time to cleanup staging data, but should be enough time to kill the Hadoop job (at least based on historical behavior). The OutputCommitter allows the user to hook into the lifecycle of the task and the jobs, so when killJob is called by the client, the AM will call OutputCommitter.abortJob() method. I have over--ridden this method, and added logic to clean up the staging data. Since the AM is in a different process than the client (running on Azkaban), the code de-serializes the WorkUnit files written to HDFS in order to obtain the writer output folder locations. Once it constructs the writer output paths, it deletes them.

@liyinan926 what do you think of this approach?

### CODE REVIEW

**Comment by liyinan926** [comment]
I think it's the right approach to rely on OutputCommitter to do the cleanup. There's an easier way to get the MR job working directory. Do the following in MRJobLauncher:

    GobblinOutputFormat.setOutputPath(mrJobDir);

Then in the OutputCommitter, you can get the path using:

    Path mrJobDir = new Path(GobblinOutputFormat.getOutputPath(context).toUri.getPath());

I also think we should put the writer staging and output directories to under the MR working directory by default, and only allow users to specify the data publisher output directory. This way we have full control to all the staging/temporary data in one place, and deleting the working directory deletes everything.

**Comment by sahilTakiar** [comment]
Apparently, NullOutputFormat doesn't have a .setOutputPath() method.

I agree with the comment on task-staging directories. Should we make the changes for moving the location of task-staging and task-output in a different pull request? It will probably require changes to AbstractJobLauncher and AvroHdfsDataWriter.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+   * functionality for cleaning up staging data when the
+   * {@link com.linkedin.uif.runtime.JobLauncher#cancelJob(Properties)} method is called via Azkaban. Azkaban only
+   * allows the cancel method run to for 5 ms until it does a hard kill on the process, so instead, we will clean up the
+   * staging data in the AM.
+   */
+  private class GobblinOutputCommitter extends OutputCommitter {
+
+    @Override
+    public void abortJob(JobContext jobContext, JobStatus.State state) throws IOException {
+      LOG.info("Aborting Job: " + jobContext.getJobID());
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: Probably also put the job state as part of the log.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+      Configuration conf = jobContext.getConfiguration();
+
+      URI fsUri = URI.create(conf.get(ConfigurationKeys.FS_URI_KEY, ConfigurationKeys.LOCAL_FS_URI));
+      FileSystem fs = FileSystem.get(fsUri, conf);
+
+      Path mrJobDir =
+          new Path(conf.get(ConfigurationKeys.MR_JOB_ROOT_DIR_KEY), conf.get(ConfigurationKeys.JOB_NAME_KEY));
+      Path jobInputDir = new Path(mrJobDir, "input");
+
+      for (FileStatus status : fs.listStatus(jobInputDir)) {
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: You can use a PathFilter to get all files that end with .wu or .mwu.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+        Closer workUnitFileCloser = Closer.create();
+        if (status.getPath().getName().endsWith(".wu")) {
+
+          WorkUnit wu = new WorkUnit();
+          wu.readFields(workUnitFileCloser.register(new DataInputStream(fs.open(status.getPath()))));
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: Please specify UTF-8 as the charset.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                  new Path(wu.getProp(ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_STAGING_DIR,
+                      i)), writerFilePath);
+
+              LOG.info("Deleting staging dir:   " + stagingDir);
+              if (wufs.exists(stagingDir)) {
+                wufs.delete(stagingDir, true);
+              }
+            }
+
+            if (wu.contains(ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_OUTPUT_DIR,
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: We should only delete the output data if the job gets cancelled.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
+package gobblin.runtime.mapreduce;
+
+import gobblin.configuration.ConfigurationKeys;
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: Put gobblin imports at the bottom.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+            Path writerFilePath =
+                new Path(ForkOperatorUtils.getPathForBranch(wu.getExtract().getOutputFilePath(), branchName,
+                    numBranches));
+
+            String uri =
+                wu.getProp(ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_FILE_SYSTEM_URI, i),
+                    ConfigurationKeys.LOCAL_FS_URI);
+
+            FileSystem wufs = FileSystem.get(URI.create(uri), new Configuration());
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: Renaming this to writeFs?

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+                      .getPropertyNameForBranch(ConfigurationKeys.WRITER_OUTPUT_DIR, i)), writerFilePath);
+
+              LOG.info("Deleting output dir:    " + outputDir);
+              if (wufs.exists(outputDir)) {
+                wufs.delete(outputDir, true);
+              }
+            }
+          }
+        }
+        // TODO add support for MultiWorkUnits
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: I think we should handle this in this pull request also.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+                new Path(ForkOperatorUtils.getPathForBranch(wu.getExtract().getOutputFilePath(), branchName,
+                    numBranches));
+
+            String uri =
+                wu.getProp(ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_FILE_SYSTEM_URI, i),
+                    ConfigurationKeys.LOCAL_FS_URI);
+
+            FileSystem wufs = FileSystem.get(URI.create(uri), new Configuration());
+
+            if (wu.contains(ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_STAGING_DIR,
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: Assign ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_STAGING_DIR, numBranches, i) to a local variable so you don't call it twice.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java (hunk 4 above)
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: Ditto.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+          wu.readFields(workUnitFileCloser.register(new DataInputStream(fs.open(status.getPath()))));
+          workUnitFileCloser.close();
+
+          LOG.info("Cleaning up staging data for WorkUnit: " + wu.getId());
+          int numBranches = wu.getPropAsInt(ConfigurationKeys.FORK_BRANCHES_KEY, 1);
+          for (int i = 0; i < numBranches; i++) {
+
+            String branchName = ForkOperatorUtils.getBranchName(wu, i, ConfigurationKeys.DEFAULT_FORK_BRANCH_NAME + i);
+
+            Path writerFilePath =
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: For the usage of methods in ForkOperatorUtils for this particular purpose, please take a look at AbstractJobLauncher.cleanupStagingData(). A task with only a single forked branch is a special case.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java (hunk 3 above)
Comment (sahilTakiar) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: I can't for a DataInputStream, also its just a serialized WorkUnit object, not a text file.

**Comment by sahilTakiar** [comment]
Addressed the comments above, updated the java docs, and wrote a test

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+      Configuration conf = jobContext.getConfiguration();
+
+      URI fsUri = URI.create(conf.get(ConfigurationKeys.FS_URI_KEY, ConfigurationKeys.LOCAL_FS_URI));
+      FileSystem fs = FileSystem.get(fsUri, conf);
+
+      Path mrJobDir =
+          new Path(conf.get(ConfigurationKeys.MR_JOB_ROOT_DIR_KEY), conf.get(ConfigurationKeys.JOB_NAME_KEY));
+      Path jobInputDir = new Path(mrJobDir, "input");
+      System.out.println("Checking: " + jobInputDir);
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: Please use LOG here.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+      // Iterate through all files in the jobInputDir, each file should correspond to a serialized wu or mwu
+      for (FileStatus status : fs.listStatus(jobInputDir, new WorkUnitFilter())) {
+
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: Additional empty line.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        Closer workUnitFileCloser = Closer.create();
+        if (status.getPath().getName().endsWith(".wu")) {
+          WorkUnit wu = new WorkUnit();
+          wu.readFields(workUnitFileCloser.register(new DataInputStream(fs.open(status.getPath()))));
+          workUnitFileCloser.close();
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: This should be called in a finally clause.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+          JobLauncherUtils.cleanStagingData(wu, LOG);
+        }
+        if (status.getPath().getName().endsWith(".mwu")) {
+          MultiWorkUnit mwu = new MultiWorkUnit();
+          mwu.readFields(workUnitFileCloser.register(new DataInputStream(fs.open(status.getPath()))));
+          workUnitFileCloser.close();
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: Ditto.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+          for (WorkUnit wu : mwu.getWorkUnits()) {
+            JobLauncherUtils.cleanStagingData(wu, LOG);
+          }
+        }
+      }
+      fs.delete(mrJobDir, true);
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: Check for existence before deleting it.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 15]
```diff
@@ -0,0 +1,131 @@
+package gobblin.runtime.mapreduce;
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: License header missing.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+      URI fsUri = URI.create(conf.get(ConfigurationKeys.FS_URI_KEY, ConfigurationKeys.LOCAL_FS_URI));
+      FileSystem fs = FileSystem.get(fsUri, conf);
+
+      Path mrJobDir =
+          new Path(conf.get(ConfigurationKeys.MR_JOB_ROOT_DIR_KEY), conf.get(ConfigurationKeys.JOB_NAME_KEY));
+      Path jobInputDir = new Path(mrJobDir, "input");
+      System.out.println("Checking: " + jobInputDir);
+      // Iterate through all files in the jobInputDir, each file should correspond to a serialized wu or mwu
+      for (FileStatus status : fs.listStatus(jobInputDir, new WorkUnitFilter())) {
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: Is it ever possible for listStatus to return a null? To be a bit more defensive, better to check for null.

File: gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+    /**
+     * Replicates the default behavior of the {@link OutputCommitter} used by {@link NullOutputFormat}.
+     */
+    public void recoverTask(TaskAttemptContext taskContext) throws IOException {
+    }
+
+    private class WorkUnitFilter implements PathFilter {
+      @Override
+      public boolean accept(Path path) {
+        return path.toString().endsWith(".wu") || path.toString().endsWith(".mwu");
```
Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/GobblinOutputFormat.java]: path.getName().endsWith().

File: gobblin-runtime/src/test/java/gobblin/runtime/mapreduce/GobblinOutputFormatTest.java [hunk 18]
```diff
@@ -0,0 +1,134 @@
+package gobblin.runtime.mapreduce;
```
Comment (liyinan926) [inline gobblin-runtime/src/test/java/gobblin/runtime/mapreduce/GobblinOutputFormatTest.java]: Missing License header.

**Comment by sahilTakiar** [comment]
Updated

**Comment by sahilTakiar** [comment]
Merged

**Pull request merged**